### ISSUE TITLE
Promise and Object.defineProperty support for Field components (for lazy properties)

### ISSUE THREAD

**Issue by rassie** [marmelab/react-admin#6175]
**Is your feature request related to a problem? Please describe.**

Depending on an API, a data provider might need to do additional requests, sometimes for each record, to fill some required fields. While this is currently possible to implement in the data provider by post-processing data, it would have to happen every time an item or a list is requested from the data provider, even if the field in question is not necessary for a particular page. 

**Describe the solution you'd like**
With `Object.defineProperty` it's possible to define getters on a record to provide additonal properties, which would be lazy-fetched. `get(record, source)` would then call the getter. Since these might make an additional API call, these will most certainly return `Promise`s. However, `react-admin`'s `Field` components currently do not resolve Promises and it also seems that properties defined with `Object.defineProperty` disappear from records after being emitted from the data provider (maybe some sanitizing routine I haven't found yet). 

**Describe alternatives you've considered**
It's also possible to define functions on a record, which can be called inside a `Field`, however, the function still wouldn't be called, but used as a value instead.

**Comment by fzaninotto** [comment]
Hi, and thanks for your suggestion.

The way we do that in react-admin is by delegating the lazy fetching to a component. Look e.g. at the `<ReferenceField>` component, which queries for related fields on mount. The advantage of doing so is that you load only what you need, it's declarative, and it can take advantage of the cache. 

I advise you to write custom Field components with an API call on mount for your lazy fields. 

And since there is already one (good) way to implement what you ask for, we won't add support for Promise values in records.

**Comment by rassie** [comment]
Thank your for the suggestion. I'm not quite sure that this approach will cover the use case I have in mind, but I'll have to try it first. Will come back to you if there is something missing.

**Comment by rassie** [comment]
@fzaninotto I've looked into it and think that my use case is not covered. If you could bear with me for just a moment, I think there is a discrepancy between common `react-admin` use case and mine. Let me give you an example.

Just assume that we have a really bad API on our hands (I'm not giving you names, but it's really bad). Let's say, there are `User`s and `Group`s in this API. A good API would look something like this: 
```
const user = { id: 123, name: "User Name", groups: [345] }
const group = { id: 345, name: "Group 1", members: [123] }
```
However, this is not the case. What we have instead is 
```
const user = { id: 123, name: "User Name" }
const group = { id: 345, name: "Group 1" }
```
and two additional API endpoints
* `/api/users/123/groups`
* `/api/groups/345/members`

So if I'm not mistaken, this is a situation in which `ReferenceField` or `ReferenceArrayField` cannot work, since there is no reference to begin with. I can certainly add a `groups` field in the data provider in post-processing, but for that I'd need to call that additional endpoint for every user record without knowing whether it's needed by a page or not.

If I understand your suggestion correctly, I'd need several application-specific `Field` components (or a wrapper for existing components), each of which would check for "special-needs" fields (e.g. `members`) and fetch additional data from the data provider, probably delegating to `react-admin`'s included `Field` components after that.

Is that correct?

**Comment by fzaninotto** [comment]
What you describe looks like a many-to-many relationship. We've built a module for that - it's called [`ra-relationships`](https://marmelab.com/ra-enterprise/modules/ra-relationships), and it's part of [react-admin Enterprise Edition](https://marmelab.com/ra-enterprise).

As we built it on react-admin, I can assure you that this use case is covered without requiring support for Promises in record fields.
